### PR TITLE
Add JAX on IPU MNIST neural net training notebook.

### PR DESCRIPTION
**Pull request by balancap**
Adapting official JAX MNIST example (using Pytorch dataloader) to run
fast on the IPU.

### CODE REVIEW

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# JAX on IPU: MNIST neural net training\n",
+    "\n",
+    "MNIST neural net training on IPU, following the original JAX tutorial: https://jax.readthedocs.io/en/latest/notebooks/Neural_Network_and_Data_Loading.html\n",
+    "\n",
+    "The original example will run as-is on the IPU. Nevertheless, to make full advantage of the IPU hardware and improve user experience, the following modifications were made:\n",
+    "\n",
+    "* `jax_platforms = \"cpu,ipu\"`: use CPU as the default platform for initialization of parameters and dataset.\n",
```
Comment (awf) [inline ipu/examples/jax_ipu_mnist_neural_net_training.ipynb:13]: Add "TODO: remove when PRNG lands"?

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "from jax.config import config\n",
+    "\n",
+    "# Use CPU as default backend for initialization.\n",
+    "config.FLAGS.jax_platforms = \"cpu,ipu\"\n",
+    "# Uncomment to use IPU model emulator.\n",
+    "config.FLAGS.jax_ipu_use_model = True\n",
+    "config.FLAGS.jax_ipu_model_num_tiles = 256"
```
Comment (awf) [inline ipu/examples/jax_ipu_mnist_neural_net_training.ipynb]: comment these

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    "  target_class = jnp.argmax(targets, axis=1)\n",
+    "  predicted_class = jnp.argmax(batched_predict(params, images), axis=1)\n",
+    "  return jnp.mean(predicted_class == target_class)\n",
+    "\n",
+    "def loss(params, images, targets):\n",
+    "  preds = batched_predict(params, images)\n",
+    "  return -jnp.mean(preds * targets)\n",
+    "\n",
+    "# Explicit jitting for IPU backend.\n",
+    "# Donate `params`` to keep variables on IPU SRAM. \n",
```
Comment (awf) [inline ipu/examples/jax_ipu_mnist_neural_net_training.ipynb]: s/variables/parameters/

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+     "shell.execute_reply": "2023-02-24T13:05:13.328033Z",
+     "shell.execute_reply.started": "2023-02-24T13:05:13.279623Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Define our dataset, using torch datasets\n",
+    "mnist_dataset = MNIST('/tmp/mnist/', download=True, transform=FlattenAndCast())\n",
+    "# IPU: dropping last micro batch with different shape.\n",
+    "training_generator = NumpyLoader(mnist_dataset, batch_size=batch_size, num_workers=0, drop_last=True)"
```
Comment (awf) [inline ipu/examples/jax_ipu_mnist_neural_net_training.ipynb:311]: note to us: remove this when doing "fast as possible" version

Comment (awf) [reply ipu/examples/jax_ipu_mnist_neural_net_training.ipynb:311]: i.e. use batch size that divides 50K

**Review by awf — APPROVED** [review]
yep lgtm.  some very minor comments.

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb (hunk 1 above)
Comment (balancap) [reply ipu/examples/jax_ipu_mnist_neural_net_training.ipynb:13]: Even when the custom IPU PRNG lands, I believe initialization of parameters will always be more user friendly on the host (due to compilation time + engine run latency). There is just not enough to do for the accelerator to really see the benefit of using it.

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb (hunk 2 above)
Comment (balancap) [reply ipu/examples/jax_ipu_mnist_neural_net_training.ipynb]: My bad, thanks!

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb (hunk 3 above)
Comment (balancap) [reply ipu/examples/jax_ipu_mnist_neural_net_training.ipynb]: Change to `parameters`.

File: ipu/examples/jax_ipu_mnist_neural_net_training.ipynb (hunk 4 above)
Comment (balancap) [reply ipu/examples/jax_ipu_mnist_neural_net_training.ipynb:311]: We can in the infeed/outfeed version. For a general notebook like that, I kind of prefer to keep to show all the pitfalls and sharp bits!

**Comment by balancap** [comment]
Thanks @awf :+1:

**Pull request merged**